### PR TITLE
#30 Color resources support

### PR DESCRIPTION
**Pull request by Tetraquark**
#30: color resources support implementation

### CODE REVIEW

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import dev.icerock.gradle.generator.android.AndroidColorsGenerator
+import dev.icerock.gradle.generator.common.CommonColorsGenerator
+import dev.icerock.gradle.generator.ios.IosColorsGenerator
+import org.gradle.api.file.FileTree
+import java.io.File
+import javax.xml.parsers.DocumentBuilderFactory
+
+abstract class ColorsGenerator(
+    private val colorsFileTree: FileTree
+) : MRGenerator.Generator, ObjectBodyExtendable {
```
Comment (Alex009) [inline gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt:22]: maybe just move `ObjectBodyExtendable` into `MRGenerator.Generator`? it duplicated in each generator class...

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        return objectBuilder.build()
+    }
+
+    override fun extendObjectBody(classBuilder: TypeSpec.Builder) = Unit
+
+    protected open fun getClassModifiers(): Array<KModifier> = emptyArray()
+    protected open fun getPropertyModifiers(): Array<KModifier> = emptyArray()
+    protected open fun getPropertyInitializer(color: ColorNode): CodeBlock? {
+        val className = if (color.isThemed()) {
+            "Themed(Color(${color.lightColor}), Color(${color.lightColor}))"
```
Comment (Alex009) [inline gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt:58]: generate named args please. this implementation can be broken if we just change order of args in runtime lib

Comment (Alex009) [reply gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt:58]: also here bug - you fill both property by lightColor

File: resources/src/androidMain/kotlin/dev/icerock/moko/resources/ColorResourceExt.kt [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * Copyright 2020 IceRock MAG Inc. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package dev.icerock.moko.resources
+
+import android.content.Context
+import android.content.res.Configuration
+import dev.icerock.moko.graphics.Color
+
+fun ColorResource.getColor(context: Context): Color {
```
Comment (Alex009) [inline resources/src/androidMain/kotlin/dev/icerock/moko/resources/ColorResourceExt.kt:11]: i think this extension required for `ColorResource.Themed` only

File: sample/android-app/src/main/java/com/icerockdev/MainActivity.kt [hunk 4]
```diff
@@ -32,6 +36,12 @@ class MainActivity : AppCompatActivity() {
         textView.text = text
         textView.typeface = Testing.getFont1().getTypeface(context = this)
 
+        val textColor = MR.colors.textColor.getColor(this).colorInt()
+            .run(Integer::toHexString)
+            .let { "#$it" }
+            .run(Color::parseColor)
```
Comment (Alex009) [inline sample/android-app/src/main/java/com/icerockdev/MainActivity.kt:42]: for what it? we got `Color` class. here for android exist `colorInt` https://github.com/icerockdev/moko-graphics/blob/master/graphics/src/androidMain/kotlin/dev/icerock/moko/graphics/ColorExt.kt#L10 that already give argb int value for setTextColor...for what here implemented this?

File: sample/ios-app/src/TestViewController.swift [hunk 5]
```diff
@@ -20,8 +20,11 @@ class TestViewController: UIViewController {
         let strings = testing.getStrings()
         
         imageView.image = drawable.toUIImage()
+        
+        let textColor = MR.colors().textColor.color
         textView.text = strings.map { $0.localized() }.joined(separator: "\n")
         textView.font = testing.getFont1().uiFont(withSize: 14.0)
+        textView.textColor = UIColor(red: CGFloat(textColor.red), green: CGFloat(textColor.green), blue: CGFloat(textColor.blue), alpha: CGFloat(textColor.alpha))
```
Comment (Alex009) [inline sample/ios-app/src/TestViewController.swift:27]: just use `textColor.toUIColor()` https://github.com/icerockdev/moko-graphics/blob/master/graphics/src/iosMain/kotlin/dev/icerock/moko/graphics/ColorExt.kt#L9

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt (hunk 1 above)
Comment (Tetraquark) [reply gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt]: Added `ObjectBodyExtendable` inheritance to `MRGenerator.Generator` interface.

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt (hunk 2 above)
Comment (Tetraquark) [reply gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt]: Added named args, fixed bug.

File: sample/android-app/src/main/java/com/icerockdev/MainActivity.kt (hunk 4 above)
Comment (Tetraquark) [reply sample/android-app/src/main/java/com/icerockdev/MainActivity.kt]: It was for tests, was removed.

File: sample/ios-app/src/TestViewController.swift (hunk 5 above)
Comment (Tetraquark) [reply sample/ios-app/src/TestViewController.swift]: Fixed.

File: resources/src/androidMain/kotlin/dev/icerock/moko/resources/ColorResourceExt.kt (hunk 3 above)
Comment (Tetraquark) [reply resources/src/androidMain/kotlin/dev/icerock/moko/resources/ColorResourceExt.kt:11]: Why? With this function an user does't need to know what color will be obtained from the resource. So you can apply a color without unnecessary checks.

File: buildSrc/src/main/kotlin/Deps.kt [hunk 6]
```diff
@@ -68,10 +68,17 @@ object Deps {
...
+                common = "dev.icerock.moko:graphics:${Versions.Libs.MultiPlatform.mokoGraphics}",
+                iosX64 = "dev.icerock.moko:graphics-iosx64:${Versions.Libs.MultiPlatform.mokoGraphics}",
+                iosArm64 = "dev.icerock.moko:graphics-iosarm64:${Versions.Libs.MultiPlatform.mokoGraphics}"
+            )
+
         }
 
         object Jvm {
             const val kotlinPoet = "com.squareup:kotlinpoet:${Versions.Libs.Jvm.kotlinPoet}"
+            const val koson = "com.lectra:koson:${Versions.Libs.Jvm.koson}"
```
Comment (Alex009) [inline buildSrc/src/main/kotlin/Deps.kt:81]: why you not use already known kotlin-serialization? for what this new library? :(

Comment (Alex009) [reply buildSrc/src/main/kotlin/Deps.kt:81]: i think we should use kotlinx-serialization

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+    override fun extendObjectBody(classBuilder: TypeSpec.Builder) = Unit
+
+    protected open fun getClassModifiers(): Array<KModifier> = emptyArray()
+    protected open fun getPropertyModifiers(): Array<KModifier> = emptyArray()
+    protected open fun getPropertyInitializer(color: ColorNode): CodeBlock? {
+        val className = if (color.isThemed()) {
+            "Themed(light = Color(0x${argbToRgba(color.lightColor)}), dark = Color(0x${argbToRgba(color.darkColor)}))"
+        } else {
+            "Single(color = Color(0x${argbToRgba(color.singleColor)}))"
```
Comment (Alex009) [inline gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt:63]: maybe in our xml use just RGBA, not ARGB? it's only in android used ARGB as i know

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/android/AndroidColorsGenerator.kt [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        valuesDir.mkdirs()
+
+        val header = """
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+            """.trimIndent()
+
+        val content = colors.map { colorNode ->
+            if (colorNode.isThemed()) {
+                "\t<color name=\"${colorNode.name}_light\">#${colorNode.lightColor}</color>\n\t<color name=\"${colorNode.name}_dark\">#${colorNode.darkColor}</color>"
```
Comment (Alex009) [inline gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/android/AndroidColorsGenerator.kt:40]: it's really correct way for themed colors? i think android support usage of one `@color/textColor` that have diferent colors for light and dark

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                arr[buildColorIdiomJsonObj(singleColor)]
+            }
+
+            val resultObj = obj {
+                "colors" to colorContentObj
+                "info" to obj {
+                    "author" to "xcode"
+                    "version" to 1
+                }
+            }
```
Comment (Alex009) [inline gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt:64]: just use kotlinx-serialization as we use in all our projects. it support this:
```
json {
    "colors" to  colorContentObj
    "info" to json {
         "author" to "xcode"
         "version" to 1
    }
}
```

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            val assetDir = File(assetsDirectory, "${colorNode.name}.colorset")
+            assetDir.mkdir()
+            val contentsFile = File(assetDir, "Contents.json")
+            contentsFile.createNewFile()
+
+            val colorContentObj = if (colorNode.isThemed()) {
+                val lightColor = parseArgbColor(colorNode.lightColor!!.toLong(16))
+                val darkColor = parseArgbColor(colorNode.darkColor!!.toLong(16))
+
+                val anyColor = buildColorIdiomJsonObj(lightColor) // set any color as light
```
Comment (Alex009) [inline gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt:47]: what it means?

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        val errors = process.errorStream.bufferedReader().readText()
+        val input = process.inputStream.bufferedReader().readText()
+        val result = process.waitFor()
+        if (result != 0) {
+            println("can't compile assets - $result")
+            println(input)
+            println(errors)
+        } else {
+            assetsDirectory.deleteRecursively()
+        }
```
Comment (Alex009) [inline gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt:82]: this part should be moved to https://github.com/icerockdev/moko-resources/blob/master/gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosMRGenerator.kt#L94
in another way feature IosImagesGenerator and IosColorsGenerator will be conflict in Assets usage

File: sample/mpp-library/src/commonMain/resources/MR/colors/colors.xml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <color name="textColor">EE2050FF</color>
+    <color name="whiteThemed">
+        <light>FFFFFFFF</light>
+        <dark>000000FF</dark>
+    </color>
```
Comment (Alex009) [inline sample/mpp-library/src/commonMain/resources/MR/colors/colors.xml:7]: i think we should test this in sample app

Comment (Alex009) [reply sample/mpp-library/src/commonMain/resources/MR/colors/colors.xml:7]: also here RGBA color. but in code you read it as ARGB...

File: resources/src/androidMain/kotlin/dev/icerock/moko/resources/ColorResourceExt.kt (hunk 3 above)
Comment (Alex009) [reply resources/src/androidMain/kotlin/dev/icerock/moko/resources/ColorResourceExt.kt:11]: agree with you

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt (hunk 10 above)
Comment (Alex009) [reply gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt:47]: https://developer.apple.com/documentation/xcode/supporting_dark_mode_in_your_interface

File: buildSrc/src/main/kotlin/Deps.kt (hunk 6 above)
Comment (Tetraquark) [reply buildSrc/src/main/kotlin/Deps.kt]: Migrated to kotlinx-serialization

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt (hunk 7 above)
Comment (Tetraquark) [reply gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ColorsGenerator.kt]: Now all colors parses as RGBA.

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/android/AndroidColorsGenerator.kt (hunk 8 above)
Comment (Tetraquark) [reply gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/android/AndroidColorsGenerator.kt]: Fixed for android dark theme.

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt (hunk 9 above)
Comment (Tetraquark) [reply gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt]: Fixed.

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt (hunk 10 above)
Comment (Tetraquark) [reply gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt]: Add comments to the code

File: gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt (hunk 11 above)
Comment (Tetraquark) [reply gradle-plugin/src/main/kotlin/dev/icerock/gradle/generator/ios/IosColorsGenerator.kt]: Fixed.

File: sample/mpp-library/src/commonMain/resources/MR/colors/colors.xml (hunk 12 above)
Comment (Tetraquark) [reply sample/mpp-library/src/commonMain/resources/MR/colors/colors.xml]: Now all colors parses as RGBA.

**Pull request merged**